### PR TITLE
Empty auth state fix

### PR DESCRIPTION
**Pull request by kunagpal**
With this change:
1. `noauth`(v2.x), `normal`(v1), and `null` are converted to null, implying that these kinds of `auth` can be overridden by default.

### CODE REVIEW

File: lib/converters/v1.0.0/converter-v1-to-v21.js [hunk 1]
```diff
@@ -52,8 +53,6 @@ _.assign(Builders.prototype, {
 
         var auth = Builders.super_.prototype.auth.call(this, entityV1);
 
-        if (auth && (auth.type === 'noauth')) { return { type: 'noauth' }; }
```
Comment (kamalaknn) [inline lib/converters/v1.0.0/converter-v1-to-v21.js]: Why is this removed?

File: lib/converters/v2.0.0/converter-v2-to-v1.js [hunk 2]
```diff
@@ -363,10 +363,11 @@ _.assign(Builders.prototype, {
                 _.isArray(v2Auth[v2Auth.type]) && (v2Auth = util.authArrayToMap(req));
                 helperAttributes = util.authMappersFromCurrent[currentHelper](v2Auth[v2Auth.type]);
             }
-            else {
-                helperAttributes = {};
-            }
         }
+        else if (v2Auth === null) {
+            currentHelper = null;
```
Comment (kamalaknn) [inline lib/converters/v2.0.0/converter-v2-to-v1.js]: Shouldn't we explicitly set `helperAttributes` to `null` as well in this case?

File: lib/normalizers/v1.js [hunk 3]
```diff
@@ -54,9 +54,11 @@ _.assign(Builders.prototype, {
         if (!entityV1) { return; }
         // Handle the noauth/normal case for new and legacy attributes first, in that order
         if (util.notLegacy(entityV1, 'auth') && entityV1.auth) {
-            return (entityV1.auth.type === 'noauth') ? { type: 'noauth' } : util.sanitizeAuthArray(entityV1);
+            return (entityV1.auth.type === 'noauth') ? null : util.sanitizeAuthArray(entityV1);
```
Comment (kamalaknn) [inline lib/normalizers/v1.js]: I think this shouldn't be the case. `normal` and `null` should be `null`. `noauth` -> `null` conversion should not be done.

File: lib/util.js [hunk 4]
```diff
@@ -113,8 +116,11 @@ util = {
             self = this,
             auth = entity && entity.auth;
 
+        if (auth === null) { return null; } // eslint-disable-line security/detect-possible-timing-attacks
         if (self.isInvalidAuth(auth)) { return; }
 
+        if (auth.type === 'noauth') { return null; }
```
Comment (kamalaknn) [inline lib/util.js]: Here as well.

File: lib/util.js [hunk 5]
```diff
@@ -140,8 +146,11 @@ util = {
...
```
Comment (kamalaknn) [inline lib/util.js]: Here too.

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 6]
```diff
@@ -242,6 +248,200 @@ describe('v1.0.0 to v2.1.0', function () {
...
+                    var source = { auth: { type: 'noauth' } };
+
+                    transformer.convertSingle(source, options, function (err, converted) {
+                        expect(err).to.not.be.ok;
+
+                        // remove `undefined` properties for testing
+                        expect(JSON.parse(JSON.stringify(converted))).to.eql({
+                            name: '',
+                            request: {
+                                auth: null,
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: This will wipe any explicit `noauth` selections on request.

File: test/unit/v1.0.0/converter-v1-to-v21.test.js [hunk 7]
```diff
@@ -242,6 +248,200 @@ describe('v1.0.0 to v2.1.0', function () {
...
+                        done();
+                    });
+                });
+            });
+        });
+
+        describe('collections', function () {
+            it('should correctly infer a noauth type from a regular auth object', function (done) {
+                var source = {
+                    auth: { type: 'noauth' },
```
Comment (kamalaknn) [inline test/unit/v1.0.0/converter-v1-to-v21.test.js]: @kunagpal even though collections will never have auth type set to `noauth` it is an allowed value. Wheather normalise should remove it is a different question.

File: lib/converters/v1.0.0/converter-v1-to-v21.js (hunk 1 above)
Comment (kunagpal) [reply lib/converters/v1.0.0/converter-v1-to-v21.js]: This is now being handled in the util call below.

File: lib/converters/v2.0.0/converter-v2-to-v1.js (hunk 2 above)
Comment (kunagpal) [reply lib/converters/v2.0.0/converter-v2-to-v1.js]: No, `helperAttributes` are not present if `currentHelper` is `null`.

File: lib/normalizers/v1.js (hunk 3 above)
Comment (kunagpal) [reply lib/normalizers/v1.js]: This is what was discussed in the previous sync up. 😓

File: lib/util.js (hunk 4 above)
Comment (kunagpal) [reply lib/util.js]: As commented above.

File: lib/util.js (hunk 5 above)
Comment (kunagpal) [reply lib/util.js]: ☝️

File: test/unit/v1.0.0/converter-v1-to-v21.test.js (hunk 7 above)
Comment (kunagpal) [reply test/unit/v1.0.0/converter-v1-to-v21.test.js]: Empty auth states are not needed for collections, this is being done to reduce the payload size.

File: lib/converters/v2.0.0/converter-v2-to-v1.js (hunk 2 above)
Comment (kamalaknn) [reply lib/converters/v2.0.0/converter-v2-to-v1.js]: Okay I remember, this is to be handled as a special case in Sync.

File: lib/normalizers/v1.js (hunk 3 above)
Comment (kamalaknn) [reply lib/normalizers/v1.js]: Consider a v2 collection with `request.auth.type='noauth'`, it gets converted to `request.auth.type='noauth'` in v1. When it gets converted back to v2 it becomes `request.auth.type=null`.

This will result in a data loss.

I'm pretty sure we did not put `noauth` in the inherit bucket in our sync up. 😟

Comment (kunagpal) [reply lib/normalizers/v1.js]: @kamalaknn As discussed offline.

File: lib/converters/v1.0.0/converter-v1-to-v21.js [hunk 8]
```diff
@@ -123,17 +126,18 @@ module.exports = {
         var auth,
             event,
             variable,
+            newCollection = {},
             units = ['info', 'item'],
             builders = new Builders(options),
-            newCollection = {},
+            authOptions = { excludeNoauth: true },
```
Comment (kamalaknn) [inline lib/converters/v1.0.0/converter-v1-to-v21.js]: Should not be needed anymore.

File: lib/converters/v2.0.0/converter-v2-to-v1.js [hunk 9]
```diff
@@ -604,9 +613,10 @@ module.exports = {
             events,
             variables,
             description,
-            units = ['order', 'folders_order', 'folders', 'requests'],
             builders = new Builders(options),
+            authOptions = { excludeNoauth: true },
```
Comment (kamalaknn) [inline lib/converters/v2.0.0/converter-v2-to-v1.js]: Same as above.

File: lib/converters/v2.1.0/converter-v21-to-v1.js [hunk 10]
```diff
@@ -16,10 +16,12 @@ _.extend(Builders.prototype, {
      * Converts arrays of v2.1 style auth params to their v1.0.0 equivalent objects.
      *
      * @param {Object} entity - A v2.1 compliant wrapped auth manifest.
+     * @param {?Object} options - The set of options for the current auth cleansing operation.
+     * @param {?Boolean} [options.includeNoauth=false] - When set to true, noauth is set to null.
```
Comment (kamalaknn) [inline lib/converters/v2.1.0/converter-v21-to-v1.js]: Which case would need `noauth` to be removed?

File: lib/normalizers/v1.js [hunk 11]
```diff
@@ -48,14 +48,18 @@ _.assign(Builders.prototype, {
...
-    auth: function (entityV1) {
+    auth: function (entityV1, options) {
         if (!entityV1) { return; }
         // Handle the noauth/normal case for new and legacy attributes first, in that order
-        if (util.notLegacy(entityV1, 'auth') && entityV1.auth) {
-            return (entityV1.auth.type === 'noauth') ? { type: 'noauth' } : util.sanitizeAuthArray(entityV1);
+        if (util.notLegacy(entityV1, 'auth')) {
+            return util.sanitizeAuthArray(entityV1, options);
         }
+        if (entityV1.currentHelper === null) { return this.options.noDefaults ? undefined : { type: 'noauth' }; }
```
Comment (kamalaknn) [inline lib/normalizers/v1.js]: As discussed in the discussion, `currentHelper=null` should not be converted to `auth.type=noauth`.

File: lib/normalizers/v1.js [hunk 12]
```diff
@@ -225,27 +229,39 @@ _.assign(Builders.prototype, {
         collectionId && !noDefaults && (requestV1.collectionId = collectionId);
 
         // normalized v1 requests should not have falsy helperAttributes or currentHelper
-        if (_.has(requestV1, 'currentHelper') && !requestV1.currentHelper) {
-            delete requestV1.currentHelper;
+        if (_.has(requestV1, 'currentHelper')) {
+            (requestV1.currentHelper === 'normal') && (requestV1.currentHelper = null);
+
+            if (!requestV1.currentHelper) {
+                (requestV1.currentHelper !== null) && (requestV1.currentHelper = null);
```
Comment (kamalaknn) [inline lib/normalizers/v1.js]: `requestV1.currentHelper = null` The only other case would be `requestV1.currentHelper === null`. It doesn't warrent to be handled explicilty.

File: lib/converters/v2.1.0/converter-v21-to-v1.js (hunk 10 above)
Comment (kunagpal) [reply lib/converters/v2.1.0/converter-v21-to-v1.js]: At a collection level.
